3단계 - 계산기록

gyurim opened the pull request:
안녕하세요, 리뷰어님 
3단계 계산기록 리뷰 요청드립니다.

감사합니다!

gyurim committed:
refactor: 리뷰 반영
- Unit 자료형 변수 네이밍 수정
- 계산식 표시 데이터 바인딩 적용

gyurim committed:
feat: data 모듈 생성 및 build.gradle 수정

gyurim committed:
feat: RoomDatabase, Dao, Entity 생성

gyurim committed:
feat: repository interface 정의 및 domain model 생성
- domain model <-> entity mapper 구현

gyurim committed:
feat: RepositoryImpl 구현

gyurim committed:
feat: 수동 DI를 위한 Injector, DataInjector 구현

gyurim committed:
feat: ViewModel 연결을 위한 ViewModelFactory 생성

gyurim committed:
feat: Record 저장, RecordList 가져오기 기능 구현

gyurim committed:
feat: CalculatorAdapter 생성 및 Adapter 적용

gyurim committed:
fix: 계산 기록 화면 수정
- 계산 기록 리스트 가져온 다음, 시계 버튼을 다시 눌렀을 때 직전 수식을 가져오도록 로직 구성
	- 수식 입력이 진행되는 와중에 시계 버튼이 눌렸다면 다시 시계 버튼이 눌렸을 때 진행하고 있었던 수식 정보를 다시 사용자에게 보여줌
- 시계 버튼이 눌렸을 때 뒤에 textview가 보이는 화면 수정
- = 버튼을 눌러 계산을 한 다음, 새로운 수식 입력을 받기 위해 기존 데이터 reset

gyurim committed:
fix: Record 모델 속 id 속성 삭제
- RepositoryImplTest에서 계산 기록 저장 및 불러오기 테스트를 구현하는 중 RecordEntity의 id 값은 autoGenerated되므로 id 값에 대한 테스트를 하지 못해 해당 속성 삭제

gyurim committed:
feat: CalculatorRepositoryImplTest 작성
- 테스트 코드 내에서 collect 메서드를 사용하니 테스트가 끝나지 않는 상황이 발생해 first() 사용

gyurim committed:
feat: CalculatorViewModel 테스트 코드 작성

gyurim commented on calculator-data/src/androidTest/java/camp/nextstep/edu/calculator/data/CalculatorRepositoryImplTest.kt:
```diff
@@ -0,0 +1,43 @@
+    fun `계산_결과_저장하고_가져오기`() {
+        val record = Record(statement = "1 + 2", result = 3)
+
+        runBlocking {
```
```kotlin
runBlocking{
    calculatorRepository.insertRecord(record)
    calculatorRepository.getAllRecords().collect{ actual->
        assertThat(actual).containsExactly(record)
    }
}
```
처음에 계산 결과 저장하고 가져오는 기능을 테스트하기 위해, 위와 같이 코드를 작성했었는데 테스트가 끝나지 않고 (틀렸다, 맞았다 결과가 나오지 않는) 계속 테스트 실행 중 상황이 발생하였습니다..! 
(actual의 결과가 null이거나 빈 배열이라면 이해가 갈텐데 테스트가 끝나지 않는 상황은 왜 발생했는지 ㅠㅠ 잘 이해가 되지 않습니다.)

```kotlin
runBlocking{
    withContext(Dispatchers.IO){calculatorRepository.insertRecord(record)}
    calculatorRepository.getAllRecords().collect{actual->
        assertThat(actual).containsExactly(record)
    }
}
```
혹시나 insert가 되기 전 모든 기록을 가지고 오는 로직을 수행하기 때문에 이와 같은 상황이 발생했나 싶어, insert가 끝난 후 기록을 가져오는 로직을 수행하도록 위와 같이 코드를 수정해보았습니다. 
그러나 테스트가 끝나지 않는 상황은 해결되지 않았습니다..! 

(물론 이 코드 외에도 중간에 emit을 직접 넣어주는 방식도 생각해보았으나 
room에서 flow를 반환해주므로, 테스트 코드 내에서 직접 emit을 해주는 것이 아닐 것 같다..라는 판단을 했습니다 ㅠㅠ)

결과적으로 하나의 값 배출을 보장하는 first()를 사용해 해결했는데, 해당 상황 해결법에 대한 조언을 얻고 싶습니다!

BeokBeok replied:
runBlocking에 대해 아래와 같이 설명이 되어 있습니다.
> Runs a new coroutine and blocks the current thread interruptibly until its completion

즉, 코루틴이 완료될때까지 스레드가 중단됩니다.
하지만 Flow는 suspend 함수가 아니기 때문에, 코루틴이 완료되었는지에 대해서 알 수 없습니다.
말씀하신 first() 함수는 종단 연산자로써, 값을 반드시 갖는 suspend 함수입니다.
해서, 코루틴이 완료되었다고 판단되어 스레드가 종료되어, 무한 대기 상태로 빠지지 않게 되는 것입니다.

Flow를 테스트하기 위해서는 간단하게 first와 같은 종단 연산자를 사용하는 방법도 있습니다만, 대체로 turbine이라는 라이브러리를 활용합니다.
ref. https://github.com/cashapp/turbine

gyurim replied:
와..! 감사합니다 👍
'suspend 함수가 아니기 때문에 코루틴이 완료되었는지에 대해 알 수 없다' 이 부분을 제가 잘 모르고 있었던 것 같아요..! 
그리고 Flow는 suspend 함수가 아니라는 것 .. 저는 flow와 suspend 함수 사이 연관성을 잘 생각하지 않고 코드를 짠 것 같아요 ㅠㅠ 피드백 정말 감사합니다

gyurim commented on calculator-app/src/test/java/camp/nextstep/edu/calculator/app/CalculatorViewModelTest.kt:
```diff
@@ -149,4 +166,59 @@ class CalculatorViewModelTest {
+        runBlocking {
+            withContext(Dispatchers.IO) {
+                calculatorViewModel.calculate()
+                calculatorRepository.insertRecord(
```
`CalculatorViewModel.calculate()` 함수 내에는 insertRecord를 하는 함수를 호출하기에, 저는 withContext 내에  `calculatorViewModel.calculate()` 코드만 작성해도 
(아래에 코드가 있습니다.)

```kotlin
runBlocking {
    withContext(Dispatchers.IO) {
        calculatorViewModel.calculate()
    }

    val actual = calculatorRepository.getAllRecords().first()
    val expected = listOf(Record("1 + 1", 2))
    assertEquals(expected, actual)
}
```
db에 record가 insert되는 로직까지 수행되는 것이 보장될 거라 생각했으나 actual에는 [] 빈 배열이 들어가 있었습니다. 
그래서 `calculatorRepository.insertRecord(record)` 코드를 넣어줘서 해결을 했습니다. 

그러나 CalculatorViewModel 속 calculate() 함수에서 `calculatorRepository.insertRecord()`가 실행되므로 테스트 코드 내에서 repository의 insert 함수가 중복 실행이 되고 있는 상황이라 생각되어 제가 해결한 방식이 좋지 못한 방식이라고 생각합니다.

이 코드를 더 좋게 작성하는 방법은 없을까요? 
그리고 왜 withContext가 calculate() 함수 속 호출된 함수의 실행 완료까지 보장을 해주지 않는지 궁금합니다.. ㅠㅠ 

--------
아래에 제가 작성한 CalculatorViewModel 속 calculate() 함수 부분을 작성해놓겠습니다! 
```Kotlin
// CalculatorViewModel.kt 
fun calculate() {
    val calculateResult = calculator.calculate(_expression.value.toString())

    if (calculateResult == null) {
        _onCalculationErrorEvent.value = Unit
    } else {
        _result.value = calculateResult!!
        _textInTextView.value = _result.value.toString()
        insertRecord()
    }

    _stateToReceiveNewInput.value = true
}

private fun insertRecord() {
    viewModelScope.launch {
        CoroutineScope(Dispatchers.IO).launch {
            calculatorRepository.insertRecord(
                Record(
                    expression.value.toString(),
                    result.value ?: 0
                )
            )
        }
    }
}
```

BeokBeok replied:
withContext로 IO Dispatcher에서 구동하도록 하였으나, calculate 함수 내에서 호출하는 insertRecord 함수는 viewModelScope에서 동작하게 되므로, 실행되지 않는 것으로 보입니다.

gyurim replied:
아하 감사합니다! 
예상하지 못한 부분이 문제였네요 🥲 
반영하여 수정하겠습니다!

BeokBeok commented on calculator-app/src/main/java/camp/nextstep/edu/calculator/CalculatorAdapter.kt:
```diff
@@ -0,0 +1,41 @@
+
+    companion object {
+        val diffUtil = object : DiffUtil.ItemCallback<Record>() {
+            override fun areItemsTheSame(oldItem: Record, newItem: Record): Boolean {
+                return oldItem == newItem
+            }
```
areItemsTheSame 함수는 객체끼리의 값을 비교하는 것이 아닌, 식별자를 비교하는 것을 권장합니다.

ref. https://developer.android.com/reference/kotlin/androidx/recyclerview/widget/DiffUtil.Callback.html#areItemsTheSame(int,int)

BeokBeok commented on calculator-app/src/main/java/camp/nextstep/edu/calculator/CalculatorViewModel.kt:
```diff
@@ -16,31 +32,108 @@ class CalculatorViewModel : ViewModel() {
+    }
+
+    private fun insertRecord() {
+        viewModelScope.launch {
+            CoroutineScope(Dispatchers.IO).launch {
+                calculatorRepository.insertRecord(
+                    Record(
+                        expression.value.toString(),
+                        result.value ?: 0
+                    )
+                )
+            }
+        }
```
viewModelScope 내에서 launch를 하는데, 그 내부에서 Dispatchers.IO로 변경하여 실행해주고 있네요.
CalculatorRepositoryImpl에서 IO Dispatcher에서 실행하도록 코드를 개선해보면 어떨까요?

gyurim replied:
네! 그렇게 반영해놓겠습니다! 

말씀 주신 코드의 장점은 
- viewmodel 속 코드 간소화 
- IO Dispatcher에서 실행한다는 로직이 viewmodel 보다는 repositoryimpl에 들어있어야 더 어울리는 로직 

이런 장점이 있다고 생각하면 맞을까요?

BeokBeok replied:
맞습니다. 👍 
이렇게 하면 ViewModel에서는 별도의 디스패쳐 스위칭 없이 일관성있게 메인스레드에서 처리해도 오류가 발생하지 않습니다.

BeokBeok reviewed (changes requested):
이전 단계 피드백과 계산기록 미션하시느라 고생하셨습니다. 👍 
고민해볼만한 의견들을 코맨트로 작성하였으니, 충분히 고민해보시고 도전해보세요. 💪 
추가로, Github Action이 실패하고 있는데 확인 부탁드려요. 🙏

gyurim committed:
fix: 리뷰 반영
- CalculatorRepositoryImpl에서 IO Dispatcher에서 insert를 실행하도록 수정
- CalculatorViewModelTest 테스트 코드 viewModelScope에서 동작하도록 수정

gyurim committed:
fix: 리뷰 반영
- Record Domain Model 속 id 속성 추가
- diffUtil areItemsTheSame()을 id 속성값으로 비교하도록 수정

gyurim committed:
fix: CalculatorViewModelTest 테스트 코드 실행 시, main thread에서 database 접근으로 인한 IllegalStateException을 Dispatchers.IO로 수정하여 해결

BeokBeok commented on calculator-app/src/main/java/camp/nextstep/edu/calculator/CalculatorAdapter.kt:
```diff
@@ -0,0 +1,41 @@
+                return oldItem.id == newItem.id
+            }
+
+            override fun areContentsTheSame(oldItem: Record, newItem: Record): Boolean {
+                return oldItem.hashCode() == newItem.hashCode()
+            }
```
```suggestion
            override fun areContentsTheSame(oldItem: Record, newItem: Record): Boolean {
                return oldItem == newItem
            }
```
areContentsTheSame은 함수명 그대로 각 객체의 컨텐츠를 비교하기 때문에, 해시코드로 비교하기 보다는 객체간 비교를 하는 것이 더 적합합니다.

BeokBeok reviewed (approved):
피드백 반영하시느라 고생하셨습니다. 👍 
추가로 고민해볼만한 의견들을 코맨트로 작성하였으니, 충분히 고민해보시고 도전해보세요. 💪 
다음 미션을 진행하시면서, 추가로 작성된 코맨트들도 반영해주세요. 🙏
